src/Stage/ReadStageInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+/*
+ *
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Stage;
+
+/**
+ * Retrieves data from the applicable data provider and sets it as a request parameter called data.
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+interface ReadStageInterface
```

Comment:
I wonder if this new concept is really necessary. Does this logic belong to the DataProvider? Or should we create a `ApiRequestFactory` class instead (see my previous comment?). I'm not sure yet, I need to think about it, but I've the feeling that we're introducing unnecessary indirection layers.

Reply:
I think their responsibility is clear:
- allow or not to apply the stage
- extract, collect and normalize the right parameters (mainly extract the identifiers and build the (de)normalization context)
- call the good "solver"
We *could* add the second point inside an `ApiRequestFactory`. But IMO it will not necessarily be a good thing. It will have a lot of conditional clauses and could quickly become a monster.
And I don't know where we could put the other parts. Maybe having a `ChainDataProvider` and modify the `ChainDataPersister` to choose the right method (`persist` or `remove` to call based on the `ApiRequest`)?

Reply:
Also IMHO, it's easier for the user to understand and decorate this kind of service than to use the data provider / persister and the serializer.

Reply:
I’m really not sure about the last point! It’s more concepts to learn with no clear responsibilities.

Reply:
I guess it depends of the developer and their way to think.
But a lot of developers are using the events already to add more logic so the need is there IMO.